extensions/pkg/controller/infrastructure/reconciler.go

```diff
@@ -218,3 +236,18 @@ func (r *reconciler) removeAnnotation(ctx context.Context, logger logr.Logger, i
 	}
 	return nil
 }
+
+func (r *reconciler) validateConfig(ctx context.Context, infrastructure *extensionsv1alpha1.Infrastructure) error {
+	if r.configValidator == nil {
+		return nil
+	}
+
+	if allErrs := r.configValidator.Validate(ctx, infrastructure); len(allErrs) > 0 {
+		if internalErrs := allErrs.Filter(field.NewErrorTypeMatcher(field.ErrorTypeInternal)); len(internalErrs) > 0 {
+			return internalErrs.ToAggregate()
+		}
+		return gardencorev1beta1helper.NewErrorWithCodes(allErrs.ToAggregate().Error(), gardencorev1beta1.ErrorConfigurationProblem)
```

Comment:
Yes, always, except if there is an internal error (e.g. a request to the cloud provider failed). I didn't think about having a different error than `ErrorConfigurationProblem `. It's perhaps a good idea, but it's a bit tricky since the validator may return multiple errors, and `field.Error` doesn't have a field to classify the error except `Type`. We could perhaps use the `Detail` field but this is a bit ugly ... Or else the `Validate` method could return a second value that is a slice of `ErrorCode`, but this would make both the interface and the `ConfigValidator` implementations a bit more complex.

I would propose to keep it simple and stick to `ErrorConfigurationProblem` for now. This could be improved later if needed.

Reply:
One approach to achieve this that wouldn't complicate the validators too much would be to add the error code to the end of the `Detail` field but only if different from `ERR_CONFIGURATION_PROBLEM`. The above function could then extract such codes (and remove it from the `Detail` field). I am still not convinced that this is needed though at this stage.